src/workflow/ArcanistBulkPatchWorkflow.php

```diff
@@ -0,0 +1,403 @@
+      }
+
+      // Run git diff repeatedly
+      foreach ($this->allDiffIds as $id) {
+        $patchfile = new TempFile();
+        $baseRef = $this->uberRefProvider->getBaseRefName($prefix, $id);
+        $diffRef = $this->uberRefProvider->getDiffRefName($prefix, $id);
+        $err = $repository_api->execPassthru('-C %s diff --binary %s %s > %s', $repoPath, $baseRef, $diffRef, $patchfile);
+        if ($err) {
+          throw new Exception(pht("Failed to generate git patch for diff %s with error code: %s", $id, $err));
+        }
+        $this->writeOkay("SUCCESS", pht('Generated patch for diff %s.', $id));
+        $diffPatchMap[$id] = $patchfile;
+      }
+    }
+    finally {
+      Filesystem::remove($repoPath);
+    }
+
+    return $diffPatchMap;
+  }
+
+  private function applyDiffWithFallback($diffId) {
+    if (empty($this->gitBasedPatches)) {
+      $bundle = $this->bundles[$diffId];
+      // Check if all patches apply
+      $err = $this->applyArcBasedPatch($bundle);
+      if ($err) {
+        $this->writeWarn(pht('PATCH FAILURE'),
+            pht('Unable to apply arc generated patch - falling back to fetching from git staging area.'));
+        $this->gitBasedPatches = $this->generateGitDiffPatches();
+        return $this->applyGitBasedPatch($diffId);
+      }
+    } else {
+      return $this->applyGitBasedPatch($diffId);
+    }
+  }
+
+
+  private function applyArcBasedPatch(ArcanistBundle $bundle) {
+    $patchfile = new TempFile();
+    Filesystem::writeFile($patchfile, $bundle->toGitPatch());
+    $passthru = new PhutilExecPassthru(
+      'git apply --whitespace nowarn --index --verbose -- %s',
+      $patchfile);
+    $passthru->setCWD($this->getRepositoryAPI()->getPath());
+    return $passthru->execute();
+  }
+
+  private function applyGitBasedPatch($diffId) {
+    $patchfile = $this->gitBasedPatches[$diffId];
+    if ($patchfile == null) {
+      $this->console->writeOut(
+        "<bg:red>** %s **</bg> %s\n",
+        pht('PATCH FAILURE'),
+        pht('Unable to find Git generated fallback patch.', $diffId));
+      return 1;
+    }
+
+    $passthru = new PhutilExecPassthru(
+      'git apply --3way --whitespace nowarn --index --verbose -- %s',
+      $patchfile);
+    $passthru->setCWD($this->getRepositoryAPI()->getPath());
+    $err = $passthru->execute();
+
+    if ($err) {
+      $this->console->writeOut(
+        "<bg:red>** %s **</bg> %s\n",
+        pht('PATCH FAILURE'),
+        pht('Unable to apply git generated patch - %s does not apply to current working state.', $diffId));
+    }
+    return $err;
+  }
+
+  // See ArcanistPatchWorkflow.php
+  private function getCommitMessage(ArcanistBundle $bundle) {
+    $revision_id    = $bundle->getRevisionID();
+    $commit_message = null;
+    $prompt_message = null;
+
+    // if we have a revision id the commit message is in differential
+
+    // TODO: See T848 for the authenticated stuff.
+    if ($revision_id && $this->isConduitAuthenticated()) {
```

Comment:
This and second conditional statement bodies are identical? Did you intend that? Doesn't seem that second conditional statement body would yield any other results

Reply:
Ok, I see that it was not very optimal PR  done by SQ team couple of years ago, we can probably just refactor it into
```
if ($revision_id) {
   $this->authenticateConduit(); // already checks if authenticated
   $conduit        = $this->getConduit();
   $commit_message = $conduit->callMethodSynchronous(
      'differential.getcommitmessage',
      array(
        'revision_id' => $revision_id,
      ));
}
```
We might even not check if revision_id is empty, if it is empty it most likely will provide a better debugging information (like json invalid or revision_id is empty from phabricator side) than error you return later when you check if commit_message is not empty

Reply:
Refactored getCommitMessage - had to keep revision_id check, as differential.getcommitmessage returns a template message if revision_id is null